src/continuous.jl

```diff
@@ -1,66 +1,96 @@
-using OrdinaryDiffEq, ForwardDiff
+#### DO NOT EDIT: CHANGING EVERYTHING TO INPLACE
+using OrdinaryDiffEq, Requires
 import OrdinaryDiffEq.ODEProblem
+import OrdinaryDiffEq.ODEIntegrator
 
-export ContinuousDS, ODEProblem
-
+export ContinuousDS, ODEProblem, ODEIntegrator
+abstract type DynamicalSystem end
 #######################################################################################
 #                                     Constructors                                    #
 #######################################################################################
 """
-    ContinuousDS(state, eom [, jacob]) <: DynamicalSystem
-`D`-dimensional continuous dynamical system (used for `D ≤ 10`).
+    ContinuousDS(state, eom! [, jacob]) <: DynamicalSystem
+`D`-dimensional continuous dynamical system.
 ## Fields:
-* `state::SVector{D}` : Current state-vector of the system, stored in the data format
+* `state::Vector{T}` : Current state-vector of the system, stored in the data format
   of `StaticArray`'s `SVector`.
-* `eom::F` (function) : The function that represents the system's equations of motion
-  (also called vector field). The function is of the format: `eom(u) -> SVector`
-  which means that given a state-vector `u` it returns an `SVector` containing the
-  derivatives `du` at the current state.
-* `jacob::J` (function) : A function that calculates the system's jacobian matrix,
-  based on the format: `jacob(u) -> SMatrix` which means that given a state-vector
-  `u` it returns an `SMatrix` containing the Jacobian at that state.
-  If the `jacob` is not provided by the user, it is created with *tremendous* efficiency
-  using the module `ForwardDiff`. Most of the time, for low dimensional systems, this
-  Jacobian is within a few % of speed of a user-defined one.
+* `eom!` (function) : The function that represents the system's equations of motion
+  (also called vector field). The function is of the format: `eom(du, u)`
+  which means that it is **in-place**, with the Julian syntax (the mutated argument
+  `du` is the first).
+* `jacob` (function) : The function that represents the Jacobian of the system,
+  given in the format: `jacob(u) => J` (i.e. returns a matrix). If the matrix is
+  an `SMatrix` from `StaticArrays.jl` there are major performance gains.
+
+Because the `jacob` function is only necessary for a small subset of algorithms, you
+do not have to provide it necessarily to the constructor (but then you can't use these
+functions).
 """
-mutable struct ContinuousDS{D, T<:Real, F, J} <: DynamicalSystem
-  state::SVector{D,T}
-  eom::F
+mutable struct ContinuousDS{T<:Real, F, J} <: DynamicalSystem
+  state::AbstractVector{T}
```

Comment:
Cool, what is the strictly typed version of it? I didn't want to use `Vector{T}`, because it could be too restrictive.

The other option would be to have 2 types in the struct definition, one the actual type of `state` and the other the `eltype(state)`.